storage/redis/peer_store.go

```diff
@@ -0,0 +1,773 @@
+}
+
+func decodePeerKey(pk serializedPeer) bittorrent.Peer {
+	peer := bittorrent.Peer{
+		ID:   bittorrent.PeerIDFromString(string(pk[:20])),
+		Port: binary.BigEndian.Uint16([]byte(pk[20:22])),
+		IP:   bittorrent.IP{IP: net.IP(pk[22:])}}
+
+	if ip := peer.IP.To4(); ip != nil {
+		peer.IP.IP = ip
+		peer.IP.AddressFamily = bittorrent.IPv4
+	} else if len(peer.IP.IP) == net.IPv6len { // implies toReturn.IP.To4() == nil
+		peer.IP.AddressFamily = bittorrent.IPv6
+	} else {
+		panic("IP is neither v4 nor v6")
+	}
+
+	return peer
+}
+
+type peerStore struct {
+	cfg Config
+	rb  *redisBackend
+
+	closed chan struct{}
+	wg     sync.WaitGroup
+}
+
+func (ps *peerStore) groups() []string {
+	return []string{bittorrent.IPv4.String(), bittorrent.IPv6.String()}
+}
+
+func (ps *peerStore) leecherInfohashKey(af, ih string) string {
+	return af + "_L_" + ih
+}
+
+func (ps *peerStore) seederInfohashKey(af, ih string) string {
+	return af + "_S_" + ih
+}
+
+func (ps *peerStore) infohashCountKey(af string) string {
+	return af + "_infohash_count"
+}
+
+func (ps *peerStore) seederCountKey(af string) string {
+	return af + "_S_count"
+}
+
+func (ps *peerStore) leecherCountKey(af string) string {
+	return af + "_L_count"
+}
+
+// populateProm aggregates metrics over all groups and then posts them to
+// prometheus.
+func (ps *peerStore) populateProm() {
+	var numInfohashes, numSeeders, numLeechers int64
+
+	conn := ps.rb.open()
+	defer conn.Close()
+
+	for _, group := range ps.groups() {
+		if n, err := conn.Do("GET", ps.infohashCountKey(group)); err != nil {
+			log.Error("storage: GET counter failure", log.Fields{
+				"key":   ps.infohashCountKey(group),
+				"error": err,
+			})
+		} else {
+			numInfohashes += n.(int64)
+		}
+		if n, err := conn.Do("GET", ps.seederCountKey(group)); err != nil {
+			log.Error("storage: GET counter failure", log.Fields{
+				"key":   ps.seederCountKey(group),
+				"error": err,
+			})
+		} else {
+			numSeeders += n.(int64)
+		}
+		if n, err := conn.Do("GET", ps.leecherCountKey(group)); err != nil {
+			log.Error("storage: GET counter failure", log.Fields{
+				"key":   ps.leecherCountKey(group),
+				"error": err,
+			})
+		} else {
+			numLeechers += n.(int64)
+		}
+	}
+
+	storage.PromInfohashesCount.Set(float64(numInfohashes))
+	storage.PromSeedersCount.Set(float64(numSeeders))
+	storage.PromLeechersCount.Set(float64(numLeechers))
+}
+
+func (ps *peerStore) getClock() int64 {
+	return timecache.NowUnixNano()
+}
+
+func (ps *peerStore) PutSeeder(ih bittorrent.InfoHash, p bittorrent.Peer) error {
+	addressFamily := p.IP.AddressFamily.String()
+	log.Debug("storage: PutSeeder", log.Fields{
+		"InfoHash": ih.String(),
+		"Peer":     p,
+	})
+
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped redis store")
+	default:
+	}
+
+	pk := newPeerKey(p)
+
+	conn := ps.rb.open()
+	defer conn.Close()
+
+	// Update the peer in the swarm.
+	encodedSeederInfoHash := ps.seederInfohashKey(addressFamily, ih.String())
```

Comment:
You can move these two lines above the `conn := ps.rb.open()` to have the connection for a shorter period of time. Not sure if that's worth anything though...